terraform/jupyterhub.tf

```diff
@@ -0,0 +1,112 @@
+resource "aws_iam_role" "ecs-service" {
+  assume_role_policy = <<EOF
+{
+    "Version": "2012-10-17",
+    "Statement": [
+        {
+            "Action": "sts:AssumeRole",
+            "Principal": {
+                "Service": "ecs.amazonaws.com"
+            },
+            "Effect": "Allow",
+            "Sid": ""
+        }
+    ]
+}
+EOF
+}
+
+resource "aws_iam_role_policy_attachment" "ecs-service" {
+  role       = "${aws_iam_role.ecs-service.name}"
+  policy_arn = "arn:aws:iam::aws:policy/AmazonEC2ContainerServiceFullAccess"
+}
+
+resource "aws_iam_instance_profile" "ecs-service" {
+  role = "${aws_iam_role.ecs-service.name}"
+}
+
+resource "aws_launch_configuration" "jupyterhub" {
+  iam_instance_profile = "${var.ecs_instance_profile}"
+  image_id             = "${var.ecs_ami}"
+  instance_type        = "m3.xlarge"
+  key_name             = "${var.key_name}"
+  security_groups      = ["${aws_security_group.security-group.id}"]
+  spot_price           = "0.05"
+
+  user_data = "#!/bin/bash\necho ECS_CLUSTER=${aws_ecs_cluster.jupyterhub.name} >> /etc/ecs/ecs.config"
+
+  lifecycle {
+    create_before_destroy = true
+  }
+}
+
+resource "aws_autoscaling_group" "jupyterhub" {
+  health_check_type    = "EC2"
+  launch_configuration = "${aws_launch_configuration.jupyterhub.id}"
+  load_balancers       = ["${aws_elb.jupyterhub.id}"]
+  vpc_zone_identifier  = ["${var.subnet}"]
+
+  desired_capacity = 1
+  max_size         = 1
+  min_size         = 1
+}
+
+resource "aws_elb" "jupyterhub" {
+  subnets         = ["${var.subnet}"]
+  security_groups = ["${aws_security_group.security-group.id}"]
+
+  listener {
+    lb_port = 8000
+    lb_protocol = "http"
+    instance_port = 8000
+    instance_protocol = "http"
+  }
+
+  health_check {
+    healthy_threshold = 3
+    unhealthy_threshold = 2
+    timeout = 3
+    target = "TCP:22"
+    interval = 5
+  }
+}
+
+data "aws_instance" "emr-master" {
+  filter {
+    name   = "dns-name"
+    values = ["${aws_emr_cluster.emr-spark-cluster.master_public_dns}"]
+  }
+}
+
+data "template_file" "jupyterhub" {
+  template = "${file("task-definitions/jupyterhub.json.template")}"
+
+  vars {
+    addr = "${data.aws_instance.emr-master.private_dns}"
+    port = "${var.jupyterhub_port}"
+  }
+}
+
+resource "aws_ecs_task_definition" "jupyterhub" {
+  container_definitions = "${data.template_file.jupyterhub.rendered}"
+  family                = "JupyterHub"
+  network_mode          = "host"
+}
+
+resource "aws_ecs_cluster" "jupyterhub" {
+  name = "JupyterHub_Cluster"
+}
+
+resource "aws_ecs_service" "jupyterhub" {
+  name            = "JupyterHub_Service"
+  cluster         = "${aws_ecs_cluster.jupyterhub.id}"
+  desired_count   = 1
+  iam_role        = "${aws_iam_role.ecs-service.name}"
+  task_definition = "${aws_ecs_task_definition.jupyterhub.arn}"
+
+  load_balancer {
```

Comment:
You weren't able to omit this? Does it work via the ELB now or is it still broken because of the WebSocket dependency?

Reply:
I have not attempted to, yet.  I would like to merge this as-is in order for the image changes to be published, then I will come back and implement your suggestions.

So far my recollection is: splitting up of the security groups and removing the `load_balancer` block.  Are there any others?

Reply:
Yeah, `security_group_rule` and trying to remove the `load_balancer` block sounds good.

Reply:
I have a work in progress PR up do deal with those two issues: https://github.com/geodocker/geodocker-jupyter-geopyspark/pull/45.